ext/pg_text_decoder.c

```diff
@@ -726,9 +743,127 @@ pg_text_dec_timestamp_without_time_zone(t_pg_coder *conv, char *val, int len, in
 	return pg_text_decoder_timestamp_do(conv, val, len, tuple, field, enc_idx, 0);
 }
 
+/*
+ * Document-class: PG::TextDecoder::Inet < PG::SimpleDecoder
+ *
+ * This is a decoder class for conversion of PostgreSQL inet type
+ * to Ruby IPAddr values.
+ *
+ */
+static VALUE
+pg_text_dec_inet(t_pg_coder *conv, char *val, int len, int tuple, int field, int enc_idx)
+{
+	VALUE ip;
+#if defined(_WIN32)
+	ip = rb_str_new(val, len);
+	ip = rb_class_new_instance(1, &ip, s_IPAddr);
+#else
+	VALUE ip_int;
+	VALUE vmasks;
+	char dst[16];
+	int af = strchr(val, '.') ? AF_INET : AF_INET6;
+	int mask = -1;
+
+	if (len >= 4) {
+		if (val[len-2] == '/') {
+			mask = val[len-1] - '0';
+			val[len-2] = '\0';
+		} else if (val[len-3] == '/') {
+			mask = (val[len-2]- '0') *10 + val[len-1] - '0';
+			val[len-3] = '\0';
+		} else if (val[len-4] == '/') {
+			mask = (val[len-3]- '0')*100 + (val[len-2]- '0')*10 + val[len-1] - '0';
+			val[len-4] = '\0';
+		}
+	}
+
+	if (1 != inet_pton(af, val, dst)) {
+		rb_raise(rb_eTypeError, "wrong data for text inet converter in tuple %d field %d val", tuple, field);
+	}
+
+	if (af == AF_INET) {
+		if (mask == -1) {
+			mask = 32;
+		} else if (mask < 0 || mask > 32) {
+			rb_raise(rb_eTypeError, "invalid mask for IPv4: %d", mask);
+		}
+		vmasks = s_vmasks4;
+
+		ip_int = UINT2NUM(ntohl(*(unsigned int *)dst));
+	} else {
+		unsigned long long * dstllp = (unsigned long long *)dst;
+
+		if (mask == -1) {
+			mask = 128;
+		} else if (mask < 0 || mask > 128) {
+			rb_raise(rb_eTypeError, "invalid mask for IPv6: %d", mask);
+		}
+		vmasks = s_vmasks6;
+
+		if (!is_big_endian) {
```

Comment:
@jeremyevans What do you think about using `read_nbo128()` [from utils.h](https://github.com/ged/ruby-pg/blob/master/ext/util.h#L44) (yet to be implemented)?
```
unsigned long long dstll = read_nbo128(dst);
ip_int = ULL2NUM(dstll);
```
This would also work if dst is non-aligned on architectures which don't allow non-alligned memory access.

Reply:
`unsigned long long` can only hold 64-bit data, not 128-bit, so I don't think that approach is possible. Using `read_nbo64` could significantly simplify the code, though.